src/pyfmi/fmi3.pyx

```diff
@@ -886,6 +1111,174 @@ cdef class FMUModelME3(FMUModelBase3):
+                xn[i] = 1.0
+            elif xn[i] < 0.0:
+                if self.callbacks.log_level >= FMIL.jm_log_level_warning:
+                    logging.warning(f"The nominal value for {xnames[i]} is <0.0 which is illegal according " + \
+                                    f"to the FMI specification. Setting the nominal to abs({xn[i]}).")
+                xn[i] = abs(xn[i])
+
+        return xn
+
+    nominal_continuous_states = property(_get_nominal_continuous_states, doc =
+    """
+    Property for accessing the nominal values of the continuous states. Calls
+    the low-level FMI function: fmi3GetNominalContinuousStates.
+    """)
+
+    cdef FMIL3.fmi3_status_t _completed_integrator_step(self,
+            FMIL3.fmi3_boolean_t no_set_FMU_state_prior_to_current_point,
+            FMIL3.fmi3_boolean_t* enter_event_mode,
+            FMIL3.fmi3_boolean_t* terminate_simulation
+        ):
+        cdef FMIL3.fmi3_status_t status
+
+        self._log_handler.capi_start_callback(self._max_log_size_msg_sent, self._current_log_size)
+        status = FMIL3.fmi3_import_completed_integrator_step(
+            self._fmu,
+            no_set_FMU_state_prior_to_current_point,
+            enter_event_mode,
+            terminate_simulation
+        )
+        self._log_handler.capi_end_callback(self._max_log_size_msg_sent, self._current_log_size)
+
+        self._last_accepted_time = self._get_time()
+        return status
+
+    def completed_integrator_step(self, no_set_FMU_state_prior_to_current_point = True):
+        """
+        This method must be called by the environment after every completed step
+        of the integrator. If returned value is True, then the environment must call
+        event_update() otherwise, no action is needed.
+
+        Returns::
+            A tuple of format (a, b) where a and b indicate:
+                If a is True -> Call event_update().
+                        False -> Do nothing.
+                If b is True -> The simulation should be terminated.
+                        False -> Do nothing.
+
+        Calls the low-level FMI function: fmi3CompletedIntegratorStep.
+        """
+        cdef FMIL3.fmi3_status_t status
+        cdef FMIL3.fmi3_boolean_t noSetFMUStatePriorToCurrentPoint = FMIL3.fmi3_true if no_set_FMU_state_prior_to_current_point else FMIL3.fmi3_false
+        cdef FMIL3.fmi3_boolean_t* enterEventMode
+        cdef FMIL3.fmi3_boolean_t* terminateSimulation
+
+        self._log_handler.capi_start_callback(self._max_log_size_msg_sent, self._current_log_size)
+        status = FMIL3.fmi3_import_completed_integrator_step(
+            self._fmu,
+            noSetFMUStatePriorToCurrentPoint,
+            enterEventMode,
+            terminateSimulation
+        )
+        self._log_handler.capi_end_callback(self._max_log_size_msg_sent, self._current_log_size)
+
+        self._last_accepted_time = self._get_time()
```

Comment:
Should this move after status check? Can we really count this as accepted if the complete_step call failed.

Also, I don't see _last_accepted_time actually being used anywhere, unless I missed something? Might as well just get rid of it?

Reply:
I have moved this now, I agree it didnt seem reasonable as it was in FMI2.